Relationship resolver version handling

RichardLake opened the pull request.

RichardLake committed:
Add comparison operators to Version.

RichardLake committed:
Change RelationshipDescriptor to use versions instead of strings.

RichardLake committed:
A utility method to RelationshipDescriptor.

RichardLake committed:
Have UniConflicts respect versions.

RichardLake committed:
Add way to pass a RelationshipDescriptor down to AvailableModule.Latest.

RichardLake committed:
Implement very basic version functionality in RelationshipResolver.

RichardLake commented:
Opps. Forgot to push the command-line changes. No not merge yet

pjf commented:
Woah! This means that when we set minimum version dependencies (and other relationships) they'll be respected? You *rock*.

RichardLake commented:
A very basic implimentation thereof but yes. 
For example:
A depends on D.
B depends on a older version of D

If the modlist to install has A,B it will throw has it can't downgrade mods. 
B,A it will work as A is happy with the old version.

Dazpoet commented:
closes #17

pjf commented on Registry/AvailableModule.cs:
```diff
@@ -43,10 +44,10 @@ public void Remove(Version version)
         ///     Returns null if there are no compatible versions.
         /// </summary>
-        public CkanModule Latest(KSPVersion ksp_version = null)
+        public CkanModule Latest(KSPVersion ksp_version = null, RelationshipDescriptor relationship=null)
```
Any chance we can get what `relationship` is used for in the method documentation?

pjf commented on Registry/AvailableModule.cs:
```diff
@@ -58,26 +59,39 @@ public CkanModule Latest(KSPVersion ksp_version = null)
+            if (ksp_version == null)
             {
-                return module_version[v];
+                var version = available_versions.FirstOrDefault(relationship.version_within_bounds);
```
So if I understand this correctly, a relationship allows us to specify a list of versions that the module can be? Such as "greater than 1.337, but less than 2.000?".

RichardLake replied:
It can either be a specific version or a minimum and/or maximum.

pjf commented on Relationships/RelationshipResolver.cs:
```diff
@@ -221,18 +221,36 @@ private void Resolve(CkanModule module, RelationshipResolverOptions options)
                 // If it's already installed, skip.
-                if (modlist.ContainsKey(dep_name) || registry.IsInstalled(dep_name))
+
+                if (modlist.ContainsKey(dep_name))
```
Note to self: This is a significant change. If we were previously treating installed module as satisfying dependencies, and now we're not, then we want to do extra testing where dependencies are installed in a separate pass, to make sure they still count in both the cmdline and the GUI clients.

It probably does make sense for the relationship resolver not to use the already installed modules, as long as whatever calls it has an easy way to add them in. If the clients themselves need to add in the list of installed mods, we should give them some sort of convenience method in the core.

pjf commented on Relationships/RelationshipResolver.cs:
```diff
@@ -221,18 +221,36 @@ private void Resolve(CkanModule module, RelationshipResolverOptions options)
+                    
+                }
+
+                if (registry.IsInstalled(dep_name))
```
Oh! Downgrade my last set of comments, I didn't see we'd split this in two.

pjf commented on Relationships/RelationshipResolver.cs:
```diff
@@ -221,18 +221,36 @@ private void Resolve(CkanModule module, RelationshipResolverOptions options)
+                    if(descriptor.version_within_bounds(registry.InstalledVersion(dep_name)))
+                        continue;
+                    //TODO Ideally we could check here if it can be replaced by the version we want. 
+                    throw new InconsistentKraken(string.Format("A certain version of {0} is needed. " +
```
For these error messages, is it easy for us to specify what that certain version is? "No" is completely fine, but I know if our users ever see these they'll want to know what version they're supposed to have installed. :)

RichardLake replied:
I'll add a helper to RelationshipDescriptor to output it and change the messages here to use it.

pjf commented on Types/Module.cs:
```diff
@@ -200,7 +221,13 @@ internal static bool UniConflicts(Module mod1, Module mod2)
+            {
+                if (!mod2.ProvidesList.Contains(conflict.name))
+                    continue;
+               if(conflict.version_within_bounds(mod2.version)) return true;
```
Okay, trying to wrap my head around the logic here...

Previously, if there was anything in `mod2` provides list that we conflicted with, we had a conflict.

Now, for each thing we conflict with...
- If it's not in mod2's provide list, we skip this (no conflict with that item)
- If it *is* in mod2's provide list, we check to see it's in our bounds, and mark it as as conflict if it is.

Hence we can conflict with "foo <= 0.05" (for example), and we'll be non-conflicting if foo 0.06 is installed.

- [ ] Reminder to pjf to check test cases for this. :)

RichardLake replied:
Yes. 
It may of been clearer if I left it in Linq. 
```mod1.conflicts.Any(conflict => mod2.ProvidesList.Contains(conflict.name))```
is changed to 
```mod1.conflicts.Any(conflict => mod2.ProvidesList.Contains(conflict.name) && conflict.version_within_bounds(mod2.version))```

pjf commented:
Written myself a bunch of line notes. It's currently 1am, and today has been a *very* hard day, but goodness me I'd like to integrate it soon. From what I've seen it looks good, but I want to go over it again with a fresh head and go over the test cases again to see if we've missed anything.

As always, @RichardLake, this is excellent work.

RichardLake committed:
Add documentation to Module.version_within_bounds. .

RichardLake committed:
Create public string RequiredVersion in Module

Helper for user friendly display of versions required to match the relationship. .

RichardLake committed:
Fix some documentation in AvailableModule.

RichardLake committed:
Alter UniConflicts to use linq.

Dazpoet commented:
How will this work with `"version" : "provided by X"` lines like the ones we see in installed-default.ckan?

Example of such a file can be seen e.g. [here](https://github.com/KSP-CKAN/CKAN/issues/911#issuecomment-104149376)
